mathesar_ui/src/components/file-attachments/file-uploader/ModalFileAttachmentUploader.svelte

```diff
@@ -0,0 +1,109 @@
+<script lang="ts">
+  import { _ } from 'svelte-i18n';
+
+  import {
+    type FileAttachmentUploadResult,
+    uploadFileAttachment,
+  } from '@mathesar/api/rest/fileAttachments';
+  import type { UploadCompletionOpts } from '@mathesar/api/rest/utils/requestUtils';
+  import { IndependentModal } from '@mathesar/component-library';
+  import ErrorBox from '@mathesar/components/message-boxes/ErrorBox.svelte';
+  import { toast } from '@mathesar/stores/toast';
+  import { getErrorMessage } from '@mathesar/utils/errors';
+  import { FileUpload as FileUploadComponent } from '@mathesar-component-library';
+  import type {
+    CancellablePromise,
+    FileUpload,
+    FileUploadAddDetail,
+  } from '@mathesar-component-library/types';
+
+  import type ModalFileAttachmentUploadController from './ModalFileAttachmentUploadController';
+
+  const MAX_DISPLAYED_COMPLETION_BEFORE_SUCCESS = 98;
+
+  export let controller: ModalFileAttachmentUploadController;
+
+  let isUploading = false;
+  let fileUploads: FileUpload[] | undefined;
+  let percentCompleted = 0;
+  let request: CancellablePromise<FileAttachmentUploadResult> | undefined =
+    undefined;
+  let error = '';
+
+  $: ({ isOpen } = controller);
+  $: firstFileUpload = fileUploads?.[0];
+  $: fileProgress = (() => {
+    if (!firstFileUpload) return {};
+    const completion = isUploading
+      ? Math.min(percentCompleted, MAX_DISPLAYED_COMPLETION_BEFORE_SUCCESS)
+      : 100;
+    return { [firstFileUpload.fileId]: completion };
+  })();
+
+  $: title = 'Upload File';
+  // I'm getting a weird JS console error when I try to use the translated title
+  // below.
+  //
+  // ```
+  // $: title = $_('upload_file');
+  // ```
+  //
+  // The error is:
+  //
+  // > "Cannot format a message without first setting the initial locale."
+  //
+  // I'm using a non-translated title for now.
+  //
+  // TODO_FILES_UI: troubleshoot and fix this error
```

Comment:
This is because this component is used within AppContext which is set before the translations are loaded. Moving `AppContext` to within the if block with `isTranslationLoading`  should ideally fix this. Just making a note here for reference.

Reply:
Ah! Thanks for pointing this out! I've linked this comment in our [task](https://internal.mathesar.org/db/36/schemas/112012/tables/112014/29).